### PR TITLE
Support code folding for argument lists

### LINKED ISSUE #67929: [Feature request]: Code folding for invocation expressions that span multiple lines
Case 1: InvocationExpressionSyntax

```csharp
var x = new C().M(
    ....,
    ....,
    ....,
    ....
);
```

Case 2: ObjectCreationExpressionSyntax

```csharp
var x = new C(
    ....,
    ....,
    ....,
    ....
);
```

Case 3: ImplicitObjectCreationExpressionSyntax

```csharp
C x = new(
    ....,
    ....,
    ....,
    ....
);
```

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #67929

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Structure/ArgumentListSyntaxStructureTests.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    [Fact]
+    public async Task TestInvocationExpressionSingleLine()
+    {
+        var code = """
+            var x = M$$();
+            """;
+
+        await VerifyBlockSpansAsync(code);
+    }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Structure/ArgumentListSyntaxStructureTests.cs:27]: test with just two args on different lines.

File: src/EditorFeatures/CSharpTest/Structure/ArgumentListSyntaxStructureTests.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    [Fact]
+    public async Task TestInvocationExpressionThreeLines()
+    {
+        var code = """
+            var x = M$${|span:(
+                "",
+                "")|};
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Structure/ArgumentListSyntaxStructureTests.cs:35]: what's the experience you have in the IDE if you have:

```c#
var v = M(N(
   "",
   ""));
```

In other words, how does nesting impact things, when the outer/inner are on the same start/end lines.

Comment (Youssef1313) [reply src/EditorFeatures/CSharpTest/Structure/ArgumentListSyntaxStructureTests.cs:35]: @CyrusNajmabadi I wasn't able to get RoslynDeployment work properly. https://www.youtube.com/watch?v=ND8dDO6oJ4U see around 38:00

**Comment by CyrusNajmabadi** [comment]
I'll try this out locally.

**Comment by CyrusNajmabadi** [comment]
So the experience is unfortunately not ok currently:

collapse

An option it to make it so that the innermost or outermost of a set of invocations with the same start line is collapsable, instead of all of htem.  

Probably the innermost makes the most sense so you still see as much code as possible.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Is it better now?

File: src/Compilers/Core/Portable/Collections/TemporaryArray`1.cs [hunk 3]
```diff
@@ -351,6 +351,58 @@ public void ReverseContents()
...
+                    }
+                    return;
+                case 3:
+                    if (compare(_item0, _item1) > 0)
+                        (_item0, _item1) = (_item1, _item0);
+
+                    if (compare(_item1, _item2) > 0)
+                        (_item1, _item2) = (_item2, _item1);
+
+                    if (compare(_item0, _item1) > 0)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Collections/TemporaryArray`1.cs:379]: >if (compare(_item0, _item1) > 0) [](http://example.com/codeflow?start=20&length=32)

It looks like there is no need for this comparison, unless we entered the previous `if`

File: src/Compilers/Core/Portable/Collections/TemporaryArray`1.cs [hunk 4]
```diff
@@ -351,6 +351,58 @@ public void ReverseContents()
...
 
+        public void Sort(Comparison<T> compare)
+        {
+            if (_builder is not null)
+            {
+                _builder.Sort(compare);
+                return;
+            }
+
+            switch (_count)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/TemporaryArray`1.cs:362]: can you use: https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.sort?view=net-8.0 instead?

File: src/Compilers/Core/CodeAnalysisTest/Collections/TemporaryArrayTests.cs [hunk 5]
```diff
@@ -205,6 +206,69 @@ public void TestReverseContents([CombinatorialRange(0, 6)] int initialItems)
                 Assert.Equal(array[i], initialItems - 1 - i);
         }
 
+        [Fact]
+        public void TestSort()
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/CodeAnalysisTest/Collections/TemporaryArrayTests.cs:210]: i would take everything out related to sorting TempArray.  We don't even relaly need temparray for block-structure.  That's an array that's good when you commonly have <4 elements.  But for block structure, we'd almost alwayshave more.  So we can just pass an ArrayBuilder along instead).

File: src/Features/Core/Portable/Structure/Syntax/AbstractBlockStructureProvider.cs [hunk 6]
```diff
@@ -38,8 +38,31 @@ public override void ProvideBlockStructure(in BlockStructureContext context)
...
+                // Then, ignore if we found a span on the same line.
+                // The effect for this is if we have something like:
+                //
+                // M1(M2(
+                //     ...
+                //     ...
+                // )
+                //
+                // We only collapse the "inner" span which has larger start.
+                spans.Sort((x, y) => y.TextSpan.Start.CompareTo(x.TextSpan.Start));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Structure/Syntax/AbstractBlockStructureProvider.cs:52]:
```suggestion
                spans.Sort(static (x, y) => y.TextSpan.Start.CompareTo(x.TextSpan.Start));
```

File: src/Compilers/Core/CodeAnalysisTest/Collections/TemporaryArrayTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/CodeAnalysisTest/Collections/TemporaryArrayTests.cs:210]: or, we can keep this.  i'm on the fence.

**Review by AlekseyTs — APPROVED** [review]
Changes under Compilers LGTM (commit 9)

File: src/Compilers/Core/Portable/Collections/TemporaryArray`1.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/TemporaryArray`1.cs:362]: Basically:

```
            Span<T> span = MemoryMarshal.CreateSpan<T>(ref _item0, this.Count);
            span.Sort(comparison);
```

?

Comment (Youssef1313) [reply src/Compilers/Core/Portable/Collections/TemporaryArray`1.cs:362]: Not available in netstandard2.0 :confused:

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Collections/TemporaryArray`1.cs:362]: Gah.  @stephentoub Any recommendations on how to best sort a span in netstandard2.0?

File: src/Compilers/Core/CodeAnalysisTest/Collections/TemporaryArrayTests.cs (hunk 5 above)
Comment (Youssef1313) [reply src/Compilers/Core/CodeAnalysisTest/Collections/TemporaryArrayTests.cs:210]: @CyrusNajmabadi I had the same feeling the TemporaryArray might not be very suitable for this scenario, but I preferred to limit the scope of my changes.

File: src/Features/Lsif/GeneratorTest/FoldingRangeTests.vb [hunk 7]
```diff
@@ -54,7 +54,7 @@ using System.Linq;|}", "imports", "...")>
                     </Workspace>, openDocuments:=False, composition:=TestLsifOutput.TestComposition)
 
                 Dim annotatedLocations = Await AbstractLanguageServerProtocolTests.GetAnnotatedLocationsAsync(workspace, workspace.CurrentSolution)
-                Dim expectedRanges = annotatedLocations("foldingRange").Select(Function(location) CreateFoldingRange(rangeKind, location.Range, collapsedText)).OrderBy(Function(range) range.StartLine).ToArray()
+                Dim expectedRanges = annotatedLocations("foldingRange").Select(Function(location) CreateFoldingRange(rangeKind, location.Range, collapsedText)).OrderByDescending(Function(range) range.StartLine).ToArray()
```
Comment (Youssef1313) [inline src/Features/Lsif/GeneratorTest/FoldingRangeTests.vb:57]: The order of the returned spans changed with the changes in this PR. I think the order shouldn't matter. So, I updated this test (and some others) to assert against the new order.

I think the tests could even ignore the order when comparing, but I didn't do that for now.

@CyrusNajmabadi Let me know your thoughts.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you take a look please? Thanks!

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Friendly reminder on this PR. Thanks!

**Comment by CyrusNajmabadi** [comment]
Yup yup. Thanks for the reminder!

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**